### ISSUE TITLE
Automatic conversion of derived classes from Eigen vector/matrix types to numpy array

### ISSUE THREAD

**Issue by allanleal** [pybind/pybind11#1338]
Consider the following class inherited from Eigen::VectorXd that prevents client code to resize it:

```c++
class VectorXdConstSize : public VectorXd
{
public:
    using VectorXd::VectorXd;

    VectorXdConstSize() = delete;

    template<typename Derived>
    auto operator=(const Eigen::MatrixBase<Derived>& other) -> VectorXdConstSize&
    {
        Ref<VectorXd> aux(*this);
        aux = other;
        return *this;
    }
    
    auto resize(Index) -> void = delete;
    
    operator Ref<VectorXd>() { return *this; }
    operator Ref<const VectorXd>() const { return *this; }
};
```

How can I get pybind11 to **automatically cast** this type to numpy array, the same way it already works for Eigen types (e.g., VectorXd, Ref<VectorXd>, etc.)?

**Comment by jagerman** [comment]
The only thing preventing this from working is the deleted default constructor: the eigen code assumes the type is default constructible, and is essentially doing:  ``type_caster<YourType> converter; /* various stuff gets checked */; converter.load(o, false);` where `o` is a python object.  The failure is in the first statement, because internally the type caster has a default-constructed `YourType value;` member.

To work around this, you'd have to duplicate the regular, dense matrix `type_caster` from eigen.h with a specialization for your class, but with a `load` that stores the internal instance in some other way (for example, inside a `std::unique_ptr<VectorXdConstSize>`, or an `std::optional` if under C++17) and extracts the instance in the appropriate way in the `operator Type` implementations.

(On a side note, with your code I think `void foo(Ref<VectorXd> v) { v.resize(5); }` *does* allow resizing).

**Comment by allanleal** [comment]
You're right - the deleted default constructor is what is preventing this. Thanks for the advice on how to overcome this.
